src/lib/utils/parsing.cpp

```diff
@@ -153,36 +153,70 @@ std::string string_join(const std::vector<std::string>& strs, char delim) {
 /*
 * Convert a decimal-dotted string to binary IP
 */
-uint32_t string_to_ipv4(std::string_view str) {
-   const auto parts = split_on(str, '.');
-
-   if(parts.size() != 4) {
-      throw Decoding_Error(fmt("Invalid IPv4 string '{}'", str));
+std::optional<uint32_t> string_to_ipv4(std::string_view str) {
+   if(str.size() > 3 + 3 * 4) {
+      return {};
    }
 
    uint32_t ip = 0;
+   size_t dots = 0;
+   uint32_t accum = 0;
+   bool dot_ok = false;
+
+   for(char c : str) {
+      if(c == '.') {
+         if(!dot_ok) {
+            return {};
+         }
+         dots += 1;
+         if(dots > 3) {
+            return {};
+         }
+
+         dot_ok = false;
+         ip = (ip << 8) | accum;
+         accum = 0;
+      } else if(c >= '0' && c <= '9') {
+         const auto d = static_cast<uint8_t>(c - '0');
 
-   for(auto part = parts.begin(); part != parts.end(); ++part) {
-      uint32_t octet = to_u32bit(*part);
+         // prohibit leading zero in quad
+         if(d == 0 && accum == 0) {
+            return {};
+         }
```

Comment:
Currently, a zero quad is not possible. E.g. something like "127.0.0.1" would fail.

Reply:
Nice catch. Added a test suite and extended the fuzzer. Also did an offline exhaustive test for all possible 32 bit addresses that string_to_ipv4 and ipv4_to_string round trip correctly.